### DISCUSSION TITLE
Get random string that's not the same as 'x'

### DISCUSSION THREAD

**Discussion by seanh20i** [bchavez/Bogus#511] [Q&A]
I'm using the Faker library to obfuscate the private information of customers. My tests simply check that the real value is not the same as the obfuscated value. For faking things like first names and states, I'm looking for a way to generate a random value that's not the same as the current value. Right now I'm using `RuleFor` to make bogus instances and I'm hoping not to have to expand all my setter lambdas into multiple lines to run multiple checks to see if subsequent results are the same. If there's an easy way to do this, I'd appreciate the knowledge...

```
        var fakeUser = new Faker<User>()
              // tests fail if random name is the same as realUser.FirstName
              .RuleFor(c => c.FirstName, f => f.Person.FirstName) 
              .Generate()
```
...would have to become:
```
        var fakeUser = new Faker<User>()
              .RuleFor(c => c.FirstName, f => {
                  var newName = f.Person.FirstName;
                  if(newName != realUser.FirstName){
                      return newName;
                  }
                  //...else run this function again until a unique string is found...
                  //This may need to be extracted from this inline lambda to support looping (?)
              })
              .Generate()
```

**Comment by dil-abertok** [comment]
Hey there,

I ran into this very same issue when generating datasets for testing negative paths. I usually need something like "here's a set of valid keys, A, B, C, I need Bogus to generate me an valid key that's not A, B, or C". What I usually end up doing is probably not an optimal solution, but it's something like this:

```c#
        public static T? GenerateUnique<T>(
            Func<T> generate,
            HashSet<T> excludedItems,
            int maxAttempts = 100)
        {
            var done = false;
            T? result = default;
            var attempt = 0;

            do
            {
                attempt++;
                var candidate = generate();

                if (!excludedItems.Contains(candidate))
                {
                    result = candidate;
                    done = true;
                }

                if (attempt >= maxAttempts)
                {
                    done = true;
                    // you might even decide to throw an exception here
                }
            } while (!done);

            return result;
        }

// usage:
var realUserNames = new Set<string> { realUser.Name };

var fakeUser = new Faker<User>()
    .RuleFor(c => c.FirstName, f => GenerateUnique(() => new Bogus.Person().FirstName, realUserNames))
    .Generate();
```

I either get a unique item in at most `maxAttempts` attempt, or get back whatever the default of my type is, meaning that I probably exhausted the pool of unique items (imagine trying to get a unique boolean value that is not true or false). How to handle that situation is pretty interesting, though, should that fail the test, or just generate a "guid name"?

Hope it helps!

**Reply by seanh20i** [reply]
Thanks for the suggestion! I did end up doing something like that with max of 3 attempts, then just returns `new Faker().Random.Word()` as it seems highly unlikely in the first place for it to choose, for example, the first name of "Eric" 3 times in a row, but also Random.Word is even more unlikely and returning the original in my case is an absolute no-go.
